### PR TITLE
Added option to reduce query fields prior to serialization

### PR DESCRIPTION
**Pull request by apmoriarty**
Query field reduction is behind a feature flag. First reduction happens at the end of query planning, second reduction happens in the VisitorFunction. This feature is also available to lookupUUID queries via the LookupUUIDTune profile.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 1]
```diff
@@ -2057,54 +2058,63 @@ protected Future<IteratorSetting> loadQueryIterator(final MetadataHelper metadat
...
-                                            QueryOptions.buildFieldStringFromSet(metadataHelper.getIndexOnlyFields(config.getDatatypeFilter())), true);
-                            addOption(cfg, QueryOptions.COMPOSITE_FIELDS,
-                                            QueryOptions.buildFieldStringFromSet(metadataHelper.getCompositeToFieldMap(config.getDatatypeFilter()).keySet()),
-                                            true);
-                            addOption(cfg, QueryOptions.INDEXED_FIELDS,
-                                            QueryOptions.buildFieldStringFromSet(metadataHelper.getIndexedFields(config.getDatatypeFilter())), true);
-                        } catch (TableNotFoundException e) {
-                            QueryException qe = new QueryException(DatawaveErrorCode.INDEX_ONLY_FIELDS_RETRIEVAL_ERROR, e);
-                            throw new DatawaveQueryException(qe);
+                            Set<String> compositeFields = metadataHelper.getCompositeToFieldMap(config.getDatatypeFilter()).keySet();
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: the refactor here clouds up the diff. Is this necessary?

**Review by FineAndDandy — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 2]
```diff
@@ -2057,54 +2058,63 @@ protected Future<IteratorSetting> loadQueryIterator(final MetadataHelper metadat
...
+                            Set<String> indexedFields = metadataHelper.getIndexedFields(config.getDatatypeFilter());
+                            Set<String> indexOnlyFields = metadataHelper.getIndexOnlyFields(config.getDatatypeFilter());
+                            
+                            // only reduce the query fields if planning has occurred
+                        if (!isPreload && config.getReduceQueryFields()) {
+                            Set<String> queryFields = ReduceFields.getQueryFields(config.getQueryTree());
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: I am very confused.  It makes perfect sense to reduce the fields in the VisitorFunction, but why reduce them here?

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java (hunk 2 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: Performing an initial intersection here reduces the number of elements intersected in the visitor function.

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: I'm looking at why this is formatting weirdly...

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java]: Fixed. No idea why that worked.

**Review by hlgp — APPROVED** [review]

**Review by alerman — APPROVED** [review]

**Review by ivakegg — APPROVED** [review]

**Pull request merged**